aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java

```diff
@@ -172,7 +172,7 @@ public synchronized void initialize(
         Objects.requireNonNull(context);
         Objects.requireNonNull(onSuccess);
         Objects.requireNonNull(onError);
-        this.threadPool = Executors.newCachedThreadPool();
+        this.threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
```

Comment:
> Reduced SQLiteStorageAdapter thread pool size to equal number of available processors

Seems like a reasonable bet for compute-bound work. But, ours is probably IO-bound. Skipping a whole lot of pedantry, I think it comes down to:

- Can we ever have more pending tasks than available processors?

Setting a cap on the thread pool sounds like a reasonable idea. Maybe the cap should be an order of magnitude higher, though.

For reference, `Schedulers.io()` in RxJava is actually _unbounded_, and has exactly the problem you called out:
> Note that this scheduler may create an unbounded number of worker threads that can result in system slowdowns or OutOfMemoryError.

Reply:
> Maybe the cap should be an order of magnitude higher, though.

My main goal in capping the thread pool is just to avoid an `OutOfMemoryError` or `CursorWindowAllocationException`, which I think we can probably accomplish by capping it to ~100.   On my Nexus 5x test device, `availableProcessors()` is 6.  I'd be okay with raising it an order of magnitude.  But why do you suggest that?  What number do you think would make sense?

Reply:
I think we're on the same page about this, but just for completeness: Software threads can be scheduled on/off of physical cores. The reason compute-bound work is usually capped to `nprocs` is because compute-bound work usually can _actually make use of_ that many cores. IO-bound work is different, it does a little work on the CPU, then waits around. Meanwhile, the kernel will schedule some other interesting work for the CPI, while the IO-bound thread sleeps. That's why IO thread pools are usually unbounded -- on average, an IO thread isn't actually doing anything and shouldn't be scheduled.

I think we probably use a handful of threads to _orchestrate_ a CRUD op on DataStore right now. What would happen if I try to save three unrelated models at once -- a `Bike`, a `HousePlant`, and a `Tuba`? We may block 6 threads before doing anything with the Tuba.

I thought maybe (nproc * 10) - (nproc * 100) would be a sweetspot. We could guard against the memory exhaustion bugs while also not serializing things too heavily.

Reply:
Thanks for the clarification!  That all makes sense to me.  I'll update it nproc * 20, if that sounds good to you!